src/gs-background-tile.c

```diff
@@ -0,0 +1,276 @@
+}
+
+static void
+gs_background_tile_setup_colors_bg (GsBackgroundTile *tile)
+{
+	GsApp *app = tile->app;
+	const guint num_colors = 4;
+	const GPtrArray *colors = gs_app_get_key_colors (app);
+	const gchar *gradients[num_colors];
+	g_autofree gchar *css = NULL;
+	gint degrees = 45;
+
+	if (colors->len == 0)
+		return;
+
+	/* apply one color gradient for each color from the cornors in the
+	 * following fashion:
+	 * -------  -------  -------  -------
+	 * -     -  - /   -  - / \ -  - / \ -
+	 * - \   -  - \   -  - \   -  - \ / -
+	 * -------  -------  -------  -------
+	 */
+	for (guint i = 0; i < num_colors; ++i) {
+		GdkRGBA *rgba = g_ptr_array_index (colors, i % colors->len);
+		guint red, green, blue;
+
+		/* we use the values in the printf below instead of using the
+		 * convenient gdk_rgba_to_string to avoid extra allocations as
+		 * this function can be used many times */
+		red = (guint) (255.0 * rgba->red);
+		green = (guint) (255.0 * rgba->green);
+		blue = (guint) (255.0 * rgba->blue);
+
+		/* the gradient will go from an solid color to the same color with more
+		 * transparency */
+		gradients[i] = g_strdup_printf ("linear-gradient(%udeg, "
+						"rgba(%u, %u, %u, 1.0), "
+						"rgba(%u, %u, %u, .2) 40%%)",
+						degrees + i * 90,
+						red, green, blue,
+						red, green, blue);
+	}
+
+	css = g_strconcat ("background: ",
+			   gradients[0], ",",
+			   gradients[1], ",",
+			   gradients[2], ",",
+			   gradients[3], ";",
+			   NULL);
+	gs_utils_widget_set_css (GTK_WIDGET (tile->image_box), css);
+
+	for (guint i = 0; i < num_colors; ++i)
+		g_free (gradients[i]);
+}
+
+static void
+gs_background_tile_update_price (GsBackgroundTile *tile)
+{
+	const GsPrice *price = NULL;
+
+	if (gs_app_is_installed (tile->app)) {
+		gtk_widget_hide (tile->price);
+		gtk_widget_show (tile->installed_icon);
+		return;
+	}
+
+	price = gs_app_get_price (tile->app);
+	if (price == NULL || (gs_price_get_amount (price) == (gdouble) 0))
+		gtk_label_set_label (GTK_LABEL (tile->price), _("FREE"));
+	else
+		gtk_label_set_label (GTK_LABEL (tile->price), gs_price_to_string (price));
+
+	gtk_widget_show (tile->price);
+	gtk_widget_hide (tile->installed_icon);
+}
+
+static gboolean
+app_state_changed_idle (gpointer user_data)
+{
+	GsBackgroundTile *tile = GS_BACKGROUND_TILE (user_data);
+	g_autofree gchar *name = NULL;
+	AtkObject *accessible = gtk_widget_get_accessible (GTK_WIDGET (tile));
+
+	if (gs_app_is_installed (tile->app))
+		name = g_strdup_printf (_("%s (Installed)"), gs_app_get_name (tile->app));
+	else if (gs_app_get_state (tile->app) == AS_APP_STATE_AVAILABLE)
+		name = g_strdup (gs_app_get_name (tile->app));
+
+	if (GTK_IS_ACCESSIBLE (accessible)) {
+		atk_object_set_name (accessible, name);
+		atk_object_set_description (accessible, gs_app_get_summary (tile->app));
+	}
+
+	if (gs_app_get_metadata_item (tile->app, "GnomeSoftware::BackgroundTile-css") == NULL)
+		gs_background_tile_setup_colors_bg (tile);
```

Comment:
Do you really need to do this again here? You do it upon setup, and if you follow my suggestion of calling it also when `css == NULL` in the function below, unless there's a bug in how the app notifies the metadata change, it should not be required here.

Reply:
This is in case we do things like refining the key-colors only after we have the tiles being shown (lazy loading) it's useful to have this.
I will add a flag to the the tile in order to skip setting the colors when they've already been added.

Reply:
My point is that if the key colors are changed, we should have received a `notify` callback already.
Right now the notify callback for this metadata key, always assumes that `css != NULL`; I think that's a bug, and if you fix the callback to call `gs_background_tile_setup_colors_bg()` when `css == NULL`, you would not need this here.
At that point you could also remove the hunk after the `/* perhaps set custom css */` comment from `gs_background_set_tile_app()` and replace it with a manual call to the notify callback.